src/pybind/mgr/cephadm/exchange.py

```diff
@@ -0,0 +1,164 @@
+# Data exchange formats for communicating more
+# complex data structures between the cephadm binary
+# an the mgr module.
+
+import json
+
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    TypeVar,
+    Union,
+    cast,
+)
+
+
+FuncT = TypeVar("FuncT", bound=Callable)
+
+
+class _DataField:
+    """A descriptor to map object fields into a data dictionary."""
+
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        field_type: Optional[FuncT] = None,
+    ):
+        self.name = name
+        self.field_type = field_type
+
+    def __set_name__(self, _: str, name: str) -> None:
+        if not self.name:
+            self.name = name
```

Comment:
is this just for logging these objects, so they log as the underlying name instead of as a `_DataField`

Reply:
The function allows the descriptor to "know" it's own name:
```python
class Foo:
   wibble = MyDesciptorType()
```
Will call `__set_name__` with args like `__set__name__(Foo, 'wibble')`.  In our _DataField type it is used as the key for the underlying data dict.